### PR TITLE
New rescan architecture + perf improvements

### PR DESCRIPTION
**Pull request by chikeichan**
This is a fairly major pull request with a new rescan architecture for Bob. The goal of the re-architecture is to fix several issues experienced by users with high lookahead values and/or lots (roughly 100+) of name that makes Bob unusable. Since this involve a fairly major refactor, I want to start with a draft first for a round of comments/feedback. All feedbacks are welcome! I would like additional scrutiny on
- `wallet/service.js` and `node/service.js`
-  edge cases around rescanning, before/after switching from node to custom rpc
- existing wallet functions ( i have tested UDPATE, REGISTER, and send/receive)

This PR involves several major changes -- I will try to be as detail as possible explaining the new changes.

**1. Removing the `hsd` hidden browser window**
If I recall correctly, the `hsd` hidden window was implemented as a workaround to make the wallet plugin works. With the `WalletNode` running as a separate service now, it seems that the hidden window is no longer necessary. This greatly simplifies the logic around `hsd` node, as well as enabling Bob to directly interface with `chaindb` for better performance (see next point for more detail).

**2. Handle requests to `hsd` and `WalletNode` using IPC**
Instead of using HTTP to make requests to both node,  Bob is interfacing with the database directly and return results to Electron using IPC calls. For wallets with large number of transactions (the one i tested on has 2000+), this improves load time from a little over a minute to under 10 seconds. 

**3. Optimize frontend load time**
Paginations is added to domain names view and my bids view.

**4. Removed `backgroundMonitor.js`**
Wallet state are fetched on view load and then cached instead of polling periodically. There is an `onNewBlock` event that each view can respond to update, but is not implemented as part of this PR.

**5. New `hsd` fork version**
I've updated `kyokan/hsd` to [this](https://github.com/kyokan/hsd/commit/98f616262e75674d5f78dfab495855824e754984) - increasing lookahead value to `uint32`. For this to be *production ready*, and new change should be added to `kyokan/hsd` to migrate `uint8` to `uint32`.

**6. Last but not least, rescan**
So here's how the rescan logic works.
1. Generate 10000 `receiveAddress` and `changeAddress` using [deriveReceive](https://github.com/handshake-org/hsd/blob/master/lib/wallet/account.js#L367) and [deriveChange](https://github.com/handshake-org/hsd/blob/master/lib/wallet/account.js#L377). Addresses are then cached so that it does not have to be derived again.
2. Using [bulk read by addresses](https://github.com/handshake-org/hsd/blob/master/lib/node/http.js#L257) to fetch all transactions from all addresses generated above, 200 addresses at a time to prevent the request from being too large. Because `hsd` index transaction by addresses, this request is almost instantaneous.
3. Using [get block header](https://github.com/handshake-org/hsd/blob/master/lib/node/http.js#L337) to fetch all block headers from 0 to current chain height. Block headers are cached so that they do not need to be fetched again after first rescan.
4. After all data are fetched, index all transactions by block number.
5. From 0 to current height, import each block + its transactions using [rescanBlock](https://github.com/handshake-org/hsd/blob/master/lib/wallet/walletdb.js#L2244)
6. Scan all related names and add them to `txdb` similar to how it's done in `hsd` [here](https://github.com/handshake-org/hsd/blob/master/lib/wallet/txdb.js#L2140)

The pro of this new architecture is that it can pair with existing `hsd` node, as I am only using methods that already existed in `hsd`. The time it takes to fetch all transactions from 10000 addresses are usually under 15 seconds as transactions are indexed by addresses. The biggest bottle neck right now come from fetching all the block headers; however, it is somewhat mitigated by caching the headers for later rescans.

I would say the happy case logic and implementation right now is fairly stable and ready for production. However, cases such as "what if Bob restarts during a rescan" is not handled yet, and will be added before next release candidate is cut and announced.

Thanks for reading all of it!

### CODE REVIEW

File: app/background/wallet/service.js [hunk 1]
```diff
@@ -152,9 +149,123 @@ class WalletService {
...
+    dispatchToMainWindow({type: START_SYNC_WALLET});
+    dispatchToMainWindow({
+      type: SYNC_WALLET_PROGRESS,
+      payload: parseInt(0),
+    });
+
+    let transactions = await this.getTXByAddresses();
+    transactions = transactions.sort((a, b) => {
+      if (a.index > b.index) return 1;
+      if (b.index > a.index) return -1;
```
Comment (pinheadmz) [inline app/background/wallet/service.js:377]: I think `index` of a TXMeta is its position in a block (for example coinbase index is `0`, etc) so I'm not sure what the advantage is to sorting here.

**Comment by pinheadmz** [comment]
This is a lot to review. I started going through it but its a bit overwhelming. I think it would be a lot easier if it were split up in to (perhaps 6) separate PRs.

I like the idea of getting rid of the hsd window. That feels like a reduction in overall complexity and we can make direct library calls on the full node instead of using HTTP. My concern there is that it will mix UI and hsd Full Node into the same process, which may have an impact on performance.

As far as the rescan/recovery stuff, I think the UI is not the right place to establish a new wallet recovery process, especially when there is an [elegant solution for the filter update issue already](https://github.com/bcoin-org/bcoin/issues/835) that just needs to be ported over to hsd, where everyone in the community can benefit from it. Using the address indexer is clever but it also occurred to me that every transaction will probably be pulled up twice (once for the sending address and again for the change address, possibly a third time for certain covenants). It also means knowing in advance what the address depth is, which we won't in a wallet recovery scenario. Setting to a super high value like 10,000 doesn't feel like the right approach, and I'm not sure what state that leaves the wallet in: first Bob will generate 20,000 address (receive and change) which could take a while on its own -- but I'm not sure it's saving those addresses back into the wallet? The user could end up with missing addresses if there were gaps in the transaction history.

Also, changing the lookahead to a u32 will require a database migration for all existing users and will break backwards compatibility if a user wants to downgrade. It also means that Bob will need to maintain its own fork of hsd in perpetuity. I know we've played with that value a lot over the past year but it doesn't feel like the right approach. We shouldn't have to pregenerate all those keys if we are just smarter about how we scan blocks at the lowest level.

**Comment by chikeichan** [comment]
> This is a lot to review. I started going through it but its a bit overwhelming. I think it would be a lot easier if it were split up in to (perhaps 6) separate PRs.

Sounds good - i can separate them into different PRs based on the 6 bullet points. I will do so before marking this PR as not a draft. I am also happy to jump on a call to walk through the changes if it helps reviewing prior to me making them into smaller PRs. The majority of this PR right now is about rescan, and most of the relevant codes (about 600 to 700 lines) are in `wallet/service.js` and `node/service.js`.

> My concern there is that it will mix UI and hsd Full Node into the same process, which may have an impact on performance.

this one should be fine -- the `hsd` full node will be running on the electron main process, where the ui will be rendered on a separate process.

> As far as the rescan/recovery stuff, I think the UI is not the right place to establish a new wallet recovery process, especially when there is an elegant solution for the filter update issue already that just needs to be ported over to hsd, where everyone in the community can benefit from it.

The new filter update is super cool!

**"UI is not the right place to establish a new recovery process"** is counter-intuitive to me. It seems more intuitive to start a new process where most of the users interactions begins. The new rescan process were also designed to work with any `hsd` node so that any one with a full `hsd` node can benefit. I do agree that the new rescan deviates from the full node implementation -- i think the new rescan when released should only be offered to users as an alternate rescan option, probably hidden in the Settings as an advanced/experimental feature.

> Using the address indexer is clever but it also occurred to me that every transaction will probably be pulled up twice (once for the sending address and again for the change address, possibly a third time for certain covenants).

I think this is okay, as the tradeoff is between _scan and filter all transactions_ vs _get transactions from index and de-duplicate_. The later still offer a significant performance upgrade.


> It also means knowing in advance what the address depth is, which we won't in a wallet recovery scenario. Setting to a super high value like 10,000 doesn't feel like the right approach, and I'm not sure what state that leaves the wallet in: first Bob will generate 20,000 address (receive and change) which could take a while on its own -- but I'm not sure it's saving those addresses back into the wallet? The user could end up with missing addresses if there were gaps in the transaction history.

Bob cached the addresses so that they only need to be derive once.

An optimization can be added to only generate 100 addresses at a time on recovery, until address index return nothing. This would mean that users MUST use address depth in sequence, otherwise quick rescan would skip addresses like you said. It seems like both options should exist -- one is great for a comprehensive first recovery, and the other is great for quick rescan when you know all of your addresses.


> Also, changing the lookahead to a u32 will require a database migration for all existing users and will break backwards compatibility if a user wants to downgrade. It also means that Bob will need to maintain its own fork of hsd in perpetuity. I know we've played with that value a lot over the past year but it doesn't feel like the right approach. We shouldn't have to pregenerate all those keys if we are just smarter about how we scan blocks at the lowest level.

Yea, I understand it would require a database migration if we flip it to u32. Based on my understanding, to avoid this we would need to implement enhanced filter in `hsd`, and then refactor the reliance on `lookahead` value inside `hsd/lib/wallet/account.js` and perhaps other places. I want to work on getting those into `hsd` eventually -- but i also want to give those users with high lookahead value a quick fix if possible. Maintaining a fork of `hsd` until enhanced filter made it into `hsd` might be our best option.

File: app/background/wallet/service.js (hunk 1 above)
Comment (chikeichan) [reply app/background/wallet/service.js:377]: It is basically pre-sorting by transactions index, before indexing them by block height. Indexing them by block height first AND THEN sort do seem more efficient though -- I will make that change.

**Review by sdtsui — COMMENTED** [review]
Hi all.  I am unable to get a dev build of this `better-scan` branch to work.

I nuked `node_nodules` and `package-lock.json` before `npm i`.  I also was:
	(i) not running a local Bob
	(ii) manually made sure any local HSD process was closed, and
	(iii) restarted my computer to make sure.  

If this is on my end or this architecture requires that I nuke my datadir first, let me know and I can try again.

Otherwise, let me know what you think.

```
sdtsuis-MacBook-Air:bob-wallet sdtsui$ npm i && npm run dev

> bob-wallet@0.6.0-rc.1 postinstall /Users/sdtsui/_dev/bob-wallet
> electron-builder install-app-deps

  • electron-builder  version=22.7.0
  • loaded configuration  file=package.json ("build" field)
  • rebuilding native dependencies  dependencies=bcrypto@5.0.4, bdb@1.2.1, bcrypto@5.3.0, bdb@1.3.0, goosig@0.9.0, mrmr@0.1.10, unbound@0.4.3 platform=darwin arch=x64
npm WARN babel-eslint@10.0.1 requires a peer of eslint@>= 4.12.1 but none is installed. You must install peer dependencies yourself.

up to date in 102.186s

55 packages are looking for funding
  run `npm fund` for details


> bob-wallet@0.6.0-rc.1 dev /Users/sdtsui/_dev/bob-wallet
> cross-env START_HOT=1 npm run start-renderer-dev


> bob-wallet@0.6.0-rc.1 start-renderer-dev /Users/sdtsui/_dev/bob-wallet
> cross-env NODE_ENV=development webpack-dev-server --config configs/webpack.config.renderer.dev.babel.js

@babel/preset-env: `DEBUG` option

Using targets:
{
  "electron": "6.0.7"
}

Using modules transform: auto

Using plugins:
  syntax-async-generators { "electron":"6.0.7" }
  syntax-object-rest-spread { "electron":"6.0.7" }
  proposal-json-strings { "electron":"6.0.7" }
  syntax-optional-catch-binding { "electron":"6.0.7" }

Using polyfills: No polyfills were added, since the `useBuiltIns` option was not set.
Starting Main Process...

> bob-wallet@0.6.0-rc.1 start-main-dev /Users/sdtsui/_dev/bob-wallet
> cross-env HOT=1 NODE_ENV=development electron --inspect=5858 -r @babel/register -r @babel/polyfill ./app/main.js

Debugger listening on ws://127.0.0.1:5858/cacf5c31-d00d-4615-bd03-73fe8bc92dee
For help, see: https://nodejs.org/en/docs/inspector
@babel/preset-env: `DEBUG` option

Using targets:
{
  "electron": "6.0.7"
}

Using modules transform: auto

Using plugins:
  syntax-async-generators { "electron":"6.0.7" }
  syntax-object-rest-spread { "electron":"6.0.7" }
  proposal-json-strings { "electron":"6.0.7" }
  syntax-optional-catch-binding { "electron":"6.0.7" }

Using polyfills: No polyfills were added, since the `useBuiltIns` option was not set.
(node:5937) UnhandledPromiseRejectionWarning: EncodingError: Out of bounds read (offset=57).
    at BufferReader.readBytes (/Users/sdtsui/_dev/bob-wallet/node_modules/bufio/lib/reader.js:713:10)
    at readKey (/Users/sdtsui/_dev/bob-wallet/node_modules/hsd/lib/wallet/account.js:860:22)
    at Account.read (/Users/sdtsui/_dev/bob-wallet/node_modules/hsd/lib/wallet/account.js:774:23)
    at Account.decode (/Users/sdtsui/_dev/bob-wallet/node_modules/bufio/lib/struct.js:94:10)
    at Function.decode (/Users/sdtsui/_dev/bob-wallet/node_modules/hsd/lib/wallet/account.js:796:26)
    at WalletDB.getAccount (/Users/sdtsui/_dev/bob-wallet/node_modules/hsd/lib/wallet/walletdb.js:1208:29)
    at Wallet.getAccount (/Users/sdtsui/_dev/bob-wallet/node_modules/hsd/lib/wallet/wallet.js:753:21)
    at Wallet.open (/Users/sdtsui/_dev/bob-wallet/node_modules/hsd/lib/wallet/wallet.js:199:21)
    at WalletDB._get (/Users/sdtsui/_dev/bob-wallet/node_modules/hsd/lib/wallet/walletdb.js:862:5)
    at WalletDB.get (/Users/sdtsui/_dev/bob-wallet/node_modules/hsd/lib/wallet/walletdb.js:830:14)
(node:5937) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 1)
(node:5937) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
(node:5937) [DEP0005] DeprecationWarning: Buffer() is deprecated due to security and usability issues. Please use the Buffer.alloc(), Buffer.allocUnsafe(), or Buffer.from() methods instead.
```

Above is my terminal output when running it for the first time. Here is the UI state - not frozen but spins indefinitely.

Screen Shot 2020-10-11 at 11 23 31 PM


If I hit refresh (CMR+R) on the electron window, I get the same error - I am pasting the full console output here:

```
app.html#/:4 Unrecognized Content-Security-Policy directive '{{cspValue}}'.

react-dom.development.js:19707 Download the React DevTools for a better development experience: https://fb.me/react-devtoolsYou might need to use a local HTTP server (instead of file://): https://fb.me/react-devtools-faq
redux-logger.js:1  action @@router/LOCATION_CHANGE @ 23:26:30.136
ipc.js:23 Dispatching IPC method call. 1 getConnection
electron/js2c/renderer_init.js:82 Electron Deprecation Warning The 'remote' module is deprecated and will be disabled by default in a future version of Electron. To ensure a smooth upgrade and silence this warning, specify {enableRemoteModule: true} in the WebPreferences for this window.
(anonymous) @ electron/js2c/renderer_init.js:82
Promise.then (async)
(anonymous) @ electron/js2c/renderer_init.js:82
./lib/renderer/api/remote.js @ electron/js2c/renderer_init.js:82
__webpack_require__ @ electron/js2c/renderer_init.js:2
loader @ electron/js2c/renderer_init.js:78
(anonymous) @ electron/js2c/renderer_init.js:30
./app/pages/DomainManager/index.js @ index.js:18
__webpack_require__ @ bootstrap:18
./app/pages/App/index.js @ index.js:1
__webpack_require__ @ bootstrap:18
./app/pages/Root.js @ Root.js:1
__webpack_require__ @ bootstrap:18
./app/index.js @ index.js:1
__webpack_require__ @ bootstrap:18
0 @ index.js:24
__webpack_require__ @ bootstrap:18
(anonymous) @ bootstrap:82
(anonymous) @ bootstrap:83
ipc.js:23 Received IPC response. 1 getConnection {type: "Custom", host: "", port: "", networkType: "", apiKey: "028b0965978137223fb9d132de96993c", …}
redux-logger.js:1  action node/START_RPC_TEST @ 23:26:30.412
ipc.js:23 Dispatching IPC method call. 2 getConnection
ipc.js:23 Received IPC response. 2 getConnection {type: "Custom", host: "", port: "", networkType: "", apiKey: "028b0965978137223fb9d132de96993c", …}
ipc.js:23 Dispatching IPC method call. 3 stop
electron/js2c/renderer_init.js:134 Electron Security Warning (Insecure Content-Security-Policy) This renderer process has either no Content Security
    Policy set or a policy with "unsafe-eval" enabled. This exposes users of
    this app to unnecessary security risks.

For more information and help, consult
https://electronjs.org/docs/tutorial/security.
This warning will not show up
once the app is packaged.
(anonymous) @ electron/js2c/renderer_init.js:134
Promise.then (async)
warnAboutInsecureCSP @ electron/js2c/renderer_init.js:134
logSecurityWarnings @ electron/js2c/renderer_init.js:134
(anonymous) @ electron/js2c/renderer_init.js:134
async function (async)
(anonymous) @ electron/js2c/renderer_init.js:134
load (async)
securityWarnings @ electron/js2c/renderer_init.js:134
(anonymous) @ electron/js2c/renderer_init.js:114
./lib/renderer/init.ts @ electron/js2c/renderer_init.js:114
__webpack_require__ @ electron/js2c/renderer_init.js:2
(anonymous) @ electron/js2c/renderer_init.js:2
___electron_webpack_init__ @ electron/js2c/renderer_init.js:2
(anonymous) @ electron/js2c/renderer_init.js:199
NativeModule.compile @ internal/bootstrap/loaders.js:287
NativeModule.compileForPublicLoader @ internal/bootstrap/loaders.js:222
loadNativeModule @ internal/modules/cjs/helpers.js:23
Module._load @ internal/modules/cjs/loader.js:698
Module._load @ electron/js2c/asar.js:769
Module.runMain @ internal/modules/cjs/loader.js:1056
(anonymous) @ internal/main/run_main_module.js:17
ipc.js:23 Received IPC response. 3 stop undefined
ipc.js:23 Dispatching IPC method call. 4 setConnectionType
ipc.js:23 Received IPC response. 4 setConnectionType undefined
ipc.js:23 Dispatching IPC method call. 5 start
ipc.js:23 Received IPC response. 5 start undefined
redux-logger.js:1  action node/STOP @ 23:26:30.520
ipc.js:23 Dispatching IPC method call. 6 get
ipc.js:23 Received IPC response. 6 get 1
ipc.js:23 Dispatching IPC method call. 7 getAccountInfo
client:52 [WDS] Live Reloading enabled.
client:126 [WDS] Warnings while compiling.
warnings @ client:126
(anonymous) @ socket.js:47
sock.onmessage @ SockJSClient.js:63
EventTarget.dispatchEvent @ sockjs.js:170
(anonymous) @ sockjs.js:887
SockJS._transportMessage @ sockjs.js:885
EventEmitter.emit @ sockjs.js:86
WebSocketTransport.ws.onmessage @ sockjs.js:2961
client:135 ./node_modules/bcrypto/lib/native/bn.js 49:17
Module parse failed: Identifier directly after number (49:17)
You may need an appropriate loader to handle this file type.
|  */
| 
> const U8_MAX = (1n << 8n) - 1n;
| const U16_MAX = (1n << 16n) - 1n;
| const U32_MAX = (1n << 32n) - 1n;
warnings @ client:135
(anonymous) @ socket.js:47
sock.onmessage @ SockJSClient.js:63
EventTarget.dispatchEvent @ sockjs.js:170
(anonymous) @ sockjs.js:887
SockJS._transportMessage @ sockjs.js:885
EventEmitter.emit @ sockjs.js:86
WebSocketTransport.ws.onmessage @ sockjs.js:2961
```

On each refresh, I get the following error in my terminal:

```
(node:5937) UnhandledPromiseRejectionWarning: Error: IO error: lock /Users/sdtsui/Library/Application Support/Electron/hsd_data/wallet/LOCK: already held by process
(node:5937) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). (rejection id: 4)
```

Does this /\ relate to the aforementioned migration work?  I haven't changed my local db or tried to import anything, just run a dev build straight from the branch.

**Comment by chikeichan** [comment]
> If this is on my end or this architecture requires that I nuke my datadir first, let me know and I can try again.

it will definitely require you to nuke your datadir first. can you try again with it removed?

**Comment by chikeichan** [comment]
closing and replacing with incremental PRs

**Pull request closed without merging**